src/v0/destinations/amazon_audience/utils.js

```diff
@@ -0,0 +1,189 @@
+  const response = defaultRequestConfig();
+  response.endpoint = '';
+  response.method = defaultPostRequestConfig.requestMethod;
+  response.headers = {
+    'Amazon-Advertising-API-ClientId': `${secret.clientId}`,
+    'Content-Type': 'application/json',
+    Authorization: `Bearer ${secret.accessToken}`,
+  };
+  response.body.JSON = {
+    createUsers: {
+      records: [
+        {
+          hashedRecords: users,
+          externalId,
+        },
+      ],
+    },
+    associateUsers: {
+      patches: [
+        {
+          op: action,
+          path: `/EXTERNAL_USER_ID-${externalId}/audiences`,
+          value: [audienceId],
+        },
+      ],
+    },
+  };
+  return response;
+};
+
+/**
+ * This function groups the response list based upon `operation`
+ * @param {*} respList
+ * @returns object
+ */
+const groupResponsesUsingOperation = (respList) => {
+  const eventGroups = lodash.groupBy(respList, (item) => item.message.action);
+  return eventGroups;
+};
+
+/**
+ * Input: [{
+    message: {
+      users: {}
+      action
+    },
+  },
+  metadata,
+  destination,
+}]
+ * @param {*} responseList 
+ */
+const batchEvents = (responseList, destination) => {
+  const { secret } = responseList[0].metadata;
+  const eventGroups = groupResponsesUsingOperation(responseList);
+  const respList = [];
+
+  Object.keys(eventGroups).forEach((op) => {
+    const { userList, jobIdList, metadataList } = eventGroups[op].reduce(
+      (acc, event) => ({
+        userList: acc.userList.concat(event.message.user),
+        jobIdList: acc.jobIdList.concat(event.metadata.jobId),
+        metadataList: acc.metadataList.concat(event.metadata),
+      }),
+      { userList: [], metadataList: [], jobIdList: [] },
+    );
+    respList.push(
+      getSuccessRespEvents(
+        buildResponseWithUsers(
+          userList,
+          op,
+          destination.config || destination.Config,
+          jobIdList,
+          secret,
+        ),
+        metadataList,
+        destination,
+        true,
+      ),
+    );
+  });
+  return respList;
+};
+
+/**
+ * This function fetches the user details and
+ * hash them after normalizing if enable hash is turned on in config
+ * @param {*} fields
+ * @param {*} config
+ * @returns
+ */
+const getUserDetails = (fields, config) => {
+  const { enableHash } = config;
+  const { email, phone, firstName, lastName, address, city, state, postalCode } = fields;
+  const user = {};
+  // formating guidelines https://advertising.amazon.com/help/GCCXMZYCK4RXWS6C
+  // using undefined as fallback so in case properties are not string
+  if (email) {
+    const em =
+      email
+        ?.replace(/[^\d.@A-Za-z-]/g, '')
+        ?.trim()
+        ?.toLowerCase() || undefined;
+    user.email = enableHash ? sha256(em) : em;
+  }
+  if (phone) {
+    const phoneNumber = phone.replace(/\D/g, '').replace(/^0+/, '');
+    user.phone = enableHash ? sha256(phoneNumber) : phoneNumber;
```

Comment:
This needs prefix with country code right? Can we use this library?
https://www.npmjs.com/package/phone